airflow/models/variable.py

```diff
@@ -143,6 +147,14 @@ def set(cls, key: str, value: Any, serialize_json: bool = False, session: Sessio
         :param serialize_json: Serialize the value to a JSON string
         :param session: SQL Alchemy Sessions
         """
+        env_var_name = "AIRFLOW_VAR_" + key.upper()
```

Comment:
Should we also do this in case we have secret backend?

Reply:
This is a bit more complicated case as this will require sending an additional request to the secret backend. The new request can be problematic in some environments as access to data in the secret is precisely monitored and audited. So even queries to check if there is an item in the backend will mean that the security department may be concerned. Especially since we don't have an API that allows us to check the existence of a key without downloading a secret.

In this change, I wanted to improve the experience of novice developers who are just starting to use Airflow and may not fully understand all the limitations. Such users suspect that they don't use the secret backend, so I don't think the lack of backend secret support is a big problem.

Reply:
Yep. I realize this is more complex. So it should be part of separate PR. And rather than querying the backend, I'd rather add somewhere an info like "origin" of a variable, so when it is read, we'd now it comes from secret backend rather than db or something like that - so that we do not increase number of queries to the DB. This should pretty simple to implement and fairly robust (I believe in all cases the variable should be read before it is written to - we might want to check all places, but it should be the case already).